iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/persistence/pipe/PipeTaskInfo.java

```diff
@@ -158,6 +162,55 @@ private void checkBeforeCreatePipeInternal(TCreatePipeReq createPipeRequest)
     throw new PipeException(exceptionMessage);
   }
 
+  public void checkBeforeAlterPipe(TAlterPipeReq alterPipeRequest) throws PipeException {
+    acquireReadLock();
+    try {
+      checkBeforeAlterPipeInternal(alterPipeRequest);
+    } finally {
+      releaseReadLock();
+    }
+  }
+
+  private void checkBeforeAlterPipeInternal(TAlterPipeReq alterPipeRequest) throws PipeException {
+    if (!isPipeExisted(alterPipeRequest.getPipeName())) {
+      final String exceptionMessage =
+          String.format(
+              "Failed to alter pipe %s, the pipe does not exist", alterPipeRequest.getPipeName());
+      LOGGER.info(exceptionMessage);
+      throw new PipeException(exceptionMessage);
+    }
+
+    PipeMeta pipeMetaFromCoordinator = getPipeMetaByPipeName(alterPipeRequest.getPipeName());
+    PipeStaticMeta pipeStaticMetaFromCoordinator = pipeMetaFromCoordinator.getStaticMeta();
+    // fill empty attributes and check useless alter from the perspective of CN
+    boolean needToAlter = false;
+    if (alterPipeRequest.getProcessorAttributes().isEmpty()) {
+      alterPipeRequest.setProcessorAttributes(
+          pipeStaticMetaFromCoordinator.getProcessorParameters().getAttribute());
+    } else if (!(new TreeMap<>(
+                pipeStaticMetaFromCoordinator.getProcessorParameters().getAttribute())
+            .toString())
+        .equals(new TreeMap<>(alterPipeRequest.getProcessorAttributes()).toString())) {
+      needToAlter = true;
+    }
+    if (alterPipeRequest.getConnectorAttributes().isEmpty()) {
+      alterPipeRequest.setConnectorAttributes(
+          pipeStaticMetaFromCoordinator.getConnectorParameters().getAttribute());
+    } else if (!(new TreeMap<>(
+                pipeStaticMetaFromCoordinator.getConnectorParameters().getAttribute())
+            .toString())
+        .equals(new TreeMap<>(alterPipeRequest.getConnectorAttributes()).toString())) {
```

Comment:
consider using keyReducer to detect identical contents like "source.xxx", "connector.xxx" or "xxx"

Reply:
To make it simple, we can ignore these cases